include/onnxruntime/core/graph/graph.h

```diff
@@ -1380,17 +1396,31 @@ class Graph {
         Graph* parent_graph,
         const Node* parent_node,
         const logging::Logger& logger,
-        bool strict_shape_type_inference);
+        bool strict_shape_type_inference,
+        std::function<void(const InlinedHashSet<std::string>&)> model_functions_remover);
```

Comment:
Do we anticipate passing different implementations of removers? If not, why have this roundabout way to remove model local functions? Doesn't the graph already have a reference to the model?

Reply:
The reason behind it is because the model reference is `const`.

Reply:
Yeah, just saw that. But now it's a bit confusing. We allow some portions of the model to be mutated even though we hold a const-ref to it?

Reply:
Yes, it is better than friendship though, and it is better than making a reference non-const.

Reply:
I get that, but the interface is confusing giving the indication that the removal function is a pluggable entity allowing different implementations when it's really done to just get around a const-ref. I didn't quite understand what's wrong with the Graph owning a non-const ref to Model. Can you please document the real rationale of passing a remover?

Reply:
It is more restrictive (than making the reference non-const), but I am not sure that it is better. Is there any concrete reason for making the reference const (other than that there was no need to modify it until now)? If making the reference non-const is possible, it seems a bit simpler to me ... my 2 cents.

Reply:
If we need to do reference counting (essentially) across the model as a whole before freeing an inlined FunctionProto, and the graph partitioner is the place that has the aggregated info on which FunctionProto instances can be removed, can we do the removal directly in GraphPartitioner::InlineFunctionsAOT instead of going via the Graph instance? The InferenceSession owns the model and can pass the non-const Model instance to InlineFunctionsAOT